File: keda/agent-worker/mcp_client.py
"""Minimal MCP-over-HTTP client, speaking the same JSON-RPC 2.0 shape that
keda/mcp-server/server.go implements (initialize / tools/list / tools/call).
No MCP SDK dependency -- this is a small, fixed protocol subset.
"""

from __future__ import annotations

import itertools
from typing import Any

import requests

_ids = itertools.count(1)


class MCPError(RuntimeError):
    pass


class MCPClient:
    def __init__(self, base_url: str, timeout: float = 15.0) -> None:
        self._url = base_url
        self._timeout = timeout
        self._session = requests.Session()

    def _call(self, method: str, params: dict[str, Any] | None = None) -> Any:
        body = {"jsonrpc": "2.0", "id": next(_ids), "method": method}
        if params is not None:
            body["params"] = params
        resp = self._session.post(self._url, json=body, timeout=self._timeout)
        resp.raise_for_status()
        data = resp.json()
        if data.get("error"):
            raise MCPError(data["error"].get("message", "unknown MCP error"))
        return data.get("result")

    def list_tools(self) -> list[dict[str, Any]]:
        """Returns tool definitions already shaped for the Anthropic Messages
        API's `tools` parameter -- MCP's `inputSchema` field maps directly to
        Claude's `input_schema`.
        """
        result = self._call("tools/list")
        return [
            {
                "name": t["name"],
                "description": t["description"],
                "input_schema": t["inputSchema"],
            }
            for t in result["tools"]
        ]

    def call_tool(self, name: str, arguments: dict[str, Any]) -> tuple[str, bool]:
        """Returns (text, is_error)."""
        result = self._call("tools/call", {"name": name, "arguments": arguments})
        text = "".join(block.get("text", "") for block in result.get("content", []))
        return text, bool(result.get("isError"))

File: keda/websocket-gateway/mcpclient.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// callTool speaks the same minimal MCP-over-HTTP JSON-RPC shape that
// keda/mcp-server implements. websocket-gateway only ever calls list_todos
// (no arguments needed), so this stays a single small function rather than
// a full client type.
func callTool(ctx context.Context, mcpURL, name string) (string, error) {
	body, err := json.Marshal(map[string]any{
		"jsonrpc": "2.0",
		"id":      1,
		"method":  "tools/call",
		"params":  map[string]any{"name": name, "arguments": map[string]any{}},
	})
	if err != nil {
		return "", err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, mcpURL, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("calling mcp-server: %w", err)
	}
	defer resp.Body.Close()

	var rpcResp struct {
		Result struct {
			Content []struct {
				Text string `json:"text"`
			} `json:"content"`
			IsError bool `json:"isError"`
		} `json:"result"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&rpcResp); err != nil {
		return "", fmt.Errorf("decoding mcp-server response: %w", err)
	}
	if rpcResp.Error != nil {
		return "", fmt.Errorf("mcp-server: %s", rpcResp.Error.Message)
	}

	var text string
	for _, c := range rpcResp.Result.Content {
		text += c.Text
	}
	if rpcResp.Result.IsError {
		return "", fmt.Errorf("mcp-server tool error: %s", text)
	}
	return text, nil
}
